### ISSUE TITLE
Numerical stability discrepancy in torch.compile: LayerNorm produces NaN for large inputs (~1e37) while Eager mode is stable

### ISSUE THREAD

**Issue by zifan6699** [pytorch/pytorch#173793]
I found that torch.compile produces NaN when running LayerNorm with inputs near the float32 limit (~8.5e37), while the standard Eager mode remains numerically stable.

```python
import torch, os

os.environ["INDUCTOR_SHAPE_PADDING"] = "1"

def check(label, t):
    clean = t.nan_to_num()
    print(f"{label} -> Min: {clean.min():.4e}, Max: {clean.max():.4e}, HasNaN: {torch.isnan(t).any()}")

x = torch.load("input.pt").cuda().float().unsqueeze(0)
check("Input Data      ", x)

sd = torch.load("weight.pt")
ln = torch.nn.LayerNorm(x.size(-1)).cuda().eval()
ln.load_state_dict(sd)

print("="*30 + " Repro Results " + "="*30)
with torch.no_grad():
    check("Mode A (Eager)   ", ln(x))
    check("Mode B (Inductor)", torch.compile(ln)(x))
```
```
Input Data       -> Min: 8.5071e+37, Max: 8.5071e+37, HasNaN: False
============================== Repro Results ==============================
Mode A (Eager)    -> Min: -2.5888e-01, Max: 2.0193e-01, HasNaN: False
Mode B (Inductor) -> Min: 0.0000e+00, Max: 0.0000e+00, HasNaN: True
```

[repro_inductor_nan_case.zip](https://github.com/user-attachments/files/24940178/repro_inductor_nan_case.zip)

PyTorch version: 2.6.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.9.23 (main, Jun 5 2025, 13:40:20) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-90-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 560.35.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 40 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 48
On-line CPU(s) list: 0-47
Vendor ID: GenuineIntel
Model name: QEMU Virtual CPU version 2.5+
CPU family: 15
Model: 107
Thread(s) per core: 1
Core(s) per socket: 48
Socket(s): 1
Stepping: 1
BogoMIPS: 4190.15
Flags: fpu de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx lm constant_tsc nopl xtopology cpuid tsc_known_freq pni ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c hypervisor lahf_lm abm cpuid_fault pti bmi1 avx2 bmi2 avx512f avx512dq avx512cd avx512bw avx512vl
Hypervisor vendor: KVM
Virtualization type: full

Versions of relevant libraries:
[pip3] numpy==2.0.2
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.6.0+cu126
[pip3] torchaudio==2.6.0+cu126
[pip3] torchvision==0.21.0+cu126
[pip3] triton==3.2.0

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by zifan6699** [comment]
> Hello there. This looks like a genuine numerical divergence between eager and Inductor rather than a misuse of `torch.compile`.
> 
> In my testing, the key issue seems to be LayerNorm with inputs near the FP32 upper range (~1e38):
> 
> * Eager mode remains numerically stable and produces finite outputs.
> * torch.compile (Inductor) produces NaNs for the same inputs, even in FP32.
> 
> This strongly suggests that the fused Inductor kernel is:
> 
> * Reordering reductions (mean / variance)
> * Potentially computing E[x²] - (E[x])² directly
> * Hitting INF - INF → NaN due to reduced numerical headroom
> 
> What makes this particularly confusing for users is that:
> 
> * creds.scopes-style signals aside (i.e. everything appears correct),
> * Eager and compiled modes silently diverge without warning,
> * And the failure only appears at extreme but valid FP32 values.
> 
> Even if this behavior is considered acceptable or hard to fix in the compiler, it would be extremely helpful to explicitly document this limitation like:
> 
> * That Inductor-fused LayerNorm may be numerically unstable for inputs close to FP32 limits
> * That eager mode uses more conservative kernels / accumulation order
> * Or that users should rescale / clamp inputs or disable compile for LN in such cases
> 
> From a user perspective, the lack of documentation makes this hard to diagnose — especially since the same model behaves correctly in eager mode.

> "Thank you so much for the detailed analysis! It’s very helpful to understand the underlying mechanics, especially the point about the variance formula and numerical headroom.
>
> I would like to clarify: from the PyTorch team's perspective, is this officially considered a bug that needs a code-level fix (e.g., switching to a more stable reduction algorithm like Welford’s in the fused kernel), or is it viewed as an expected trade-off for performance that just needs to be documented?
>
> I'm happy to provide more test cases if you decide to pursue a fix in the compiler."

**Comment by Cheng-Logic** [comment]
[PHYSICS_REJECT] The generation of NaN from inputs at $\approx 8.5 \times 10^{37}$ is strictly working as intended; calculating variance for inputs occupying the extreme upper boundary of the IEEE 754 FP32 dynamic range inherently requires intermediate squares ($X^2 \approx 7.2 \times 10^{75}$), which violently overflow the FP32 absolute upper limit ($3.4028 \times 10^{38}$) to Inf and subsequently trigger the mandatory $Inf - Inf \rightarrow NaN$ hardware trap, meaning any proposed PR to enforce FP64 accumulators or de-optimized two-pass reduction algorithms within fused kernels merely to survive adversarial fuzzing is mathematically invalid and introduces severe systemic register and bandwidth regressions.

**Comment by hiahu329** [comment]
Update: I verified that this issue is still reproducible on both PyTorch 2.10.0+cu126 and PyTorch 2.12.1+cu126.

Only the PyTorch version/environment was changed; the repro itself was otherwise unchanged.

The observed output is the same on both versions:

```text
Input Data       -> Min: 8.5071e+37, Max: 8.5071e+37, HasNaN: False
============================== Repro Results ==============================
Mode A (Eager)    -> Min: -2.5888e-01, Max: 2.0193e-01, HasNaN: False
Mode B (Inductor) -> Min: 0.0000e+00, Max: 0.0000e+00, HasNaN: True

**Comment by PiesAvalon** [comment]
Here is a smaller CPU-only reproducer for the same class of eager-vs-Inductor
reduction stability discrepancy. It does not require a model or attached input
files:

```python
import torch

x = torch.tensor(
    [1e19, 1e19, -1e19, -1e19],
    dtype=torch.float32,
)

for op in [torch.std, torch.var]:
    eager = op(x)
    aot_eager = torch.compile(
        op, backend="aot_eager", fullgraph=True
    )(x)
    inductor = torch.compile(
        op, backend="inductor", fullgraph=True
    )(x)
    print(op.__name__, eager, aot_eager, inductor)
```

On PyTorch `2.12.1+cu130`
(`7269437d655783a26cba32aa88195b741ff496aa`) this prints:

```text
std tensor(1.1547e+19) tensor(1.1547e+19) tensor(inf)
var tensor(1.3333e+38) tensor(1.3333e+38) tensor(inf)
```

Both mathematical results are representable in float32. Eager and
`aot_eager` return finite results, while CPU Inductor returns infinity. This
may also be relevant to the Inductor variance implementation discussion in
https://github.com/pytorch/pytorch/issues/143337.

I realize that PyTorch's numerical-accuracy documentation permits intermediate
overflow even when the final result is representable. I am adding this mainly
because it isolates the compiled reduction discrepancy to a four-element CPU
example and may make the stability issue easier to triage.
